src/main/java/com/microsoft/azure/vmagent/util/AzureClientFactory.java

```diff
@@ -0,0 +1,131 @@
+
+    public static String getUserAgent() {
+        String version = null;
+        String instanceId = null;
+        try {
+            version = AzureClientFactory.class.getPackage().getImplementationVersion();
+            instanceId = Jenkins.getInstance().getLegacyInstanceId();
+        } catch (Exception e) {
+        }
+
+        if (version == null) {
+            version = "local";
+        }
+        if (instanceId == null) {
+            instanceId = "local";
+        }
+
+        return Constants.PLUGIN_NAME + "/" + version + "/" + instanceId;
+    }
+
+    @Nonnull
+    public static Azure getClient(String credentialId) {
+        AzureMsiCredentials credential = CredentialsMatchers.firstOrNull(
+                CredentialsProvider.lookupCredentials(
+                        AzureMsiCredentials.class,
+                        Jenkins.getInstance(),
+                        ACL.SYSTEM,
+                        Collections.<DomainRequirement>emptyList()),
+                CredentialsMatchers.withId(credentialId));
+        if (credential != null) {
+            try {
+                return getClient(credential.getMsiPort());
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        } else {
+            AzureCredentials.ServicePrincipal servicePrincipal = AzureCredentials.getServicePrincipal(credentialId);
+            return getClient(servicePrincipal);
+        }
+    }
+
+    @Nonnull
+    public static Azure getClient(final AzureCredentials.ServicePrincipal credentials) {
+        if (credentials == null) {
+            throw new NullPointerException();
+        }
+        ApplicationTokenCredentials token = new ApplicationTokenCredentials(
+                credentials.getClientId(),
+                credentials.getTenant(),
+                credentials.getClientSecret(),
+                new AzureEnvironment(ImmutableMap.of(
+                        "activeDirectoryEndpointUrl", credentials.getAuthenticationEndpoint(),
+                        "activeDirectoryGraphResourceId", credentials.getGraphEndpoint(),
+                        "managementEndpointUrl", credentials.getServiceManagementURL(),
+                        "resourceManagerEndpointUrl", credentials.getResourceManagerEndpoint(),
+                        "activeDirectoryResourceId", "https://management.core.windows.net/")));
+        return Azure.configure()
+                .withInterceptor(new AzureVMAgentPlugin.AzureTelemetryInterceptor())
+                .withLogLevel(Constants.DEFAULT_AZURE_SDK_LOGGING_LEVEL)
+                .withUserAgent(getUserAgent())
+                .authenticate(token)
+                .withSubscription(credentials.getSubscriptionId());
+    }
+
+    @Nonnull
+    public static Azure getClient(int msiPort) throws IOException {
+        MsiTokenCredentials msiToken = new MsiTokenCredentials(msiPort, AzureEnvironment.AZURE);
```

Comment:
Does this work for other Azure environment?

Reply:
currently not. I'm adding it in the credentials plugin